### PR TITLE
Peptides

### LINKED ISSUE #452: Import Evidence File
## Issue description
- In Section Importing, Step Peptide Import write a Method "Import Evidence", which takes a path to an Evidence file and imports the relevant Data
- The Relevant Columns for now are (according to Chris): 
Sequence, Modification, modified sequence, Proteins, Missed cleavages, Raw file / experiment, Intensity, PEP

## Acceptance Criteria
- [x] an Import Evidence Method exists and can be selected by the user
- [x] the method returns a df with all the relevant columns (for now those are Sequence, Modification, modified sequence, Proteins, Missed cleavages, Raw file / experiment, Intensity, PEP)

### PR DESCRIPTION
**Pull request by JanniRoebbecke**
## Description
fixes #452, #453, #454, #455, #456, #461

- Peptide Data can now be imported and is then tagged along through the preprocessing (in every preprocessing step the peptide data is modified to match the peptide data) 
- (see the linked issues above to check how the different preprocessing steps should handle peptide data, as described by Chris)

## Changes

- all preprocessing methods were adapted to take a peptide dataframe, modify and return it, matching the peptide dataframe
- all preprocessing tests were adapted to test the new functionality
- the step classes were addapted to hand the peptide data from one preprocessing step to the next (if peptide data was imported)

- the step evidence import was added to support the format of the evidence file (including tests for this new method and the necessary forms and step class)

## Testing

- step trough the run and check the peptide dataframes in the output trough brakepoints

## PR checklist
**Development**
- [x] If necessary, I have updated the documentation (README, docstrings, etc.)
- [x] If necessary, I have created / updated tests.
 
**Mergeability**
- [x] main-branch has been merged into local branch to resolve conflicts
- [ ] The tests and linter have passed AFTER local merge
- [ ] The code has been formatted with `black`
 
**Code review**
- [x] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

### CODE REVIEW

File: protzilla/data_preprocessing/normalisation.py [hunk 1]
```diff
@@ -171,7 +172,8 @@ def by_totalsum(protein_df: pd.DataFrame) -> dict:
 
 def by_reference_protein(
     protein_df: pd.DataFrame,
-    reference_protein: str,
+    peptide_df: pd.DataFrame = None,
+    reference_protein: str = "",
```
[the remark this answers is not in the document]
Comment (JanniRoebbecke) [reply protzilla/data_preprocessing/normalisation.py]: removed default values

File: protzilla/data_preprocessing/filter_samples.py [hunk 2]
```diff
@@ -5,12 +5,15 @@
 
 
 def by_protein_intensity_sum(
-    protein_df: pd.DataFrame, deviation_threshold: float
+    protein_df: pd.DataFrame = None,
```
[the remark this answers is not in the document]
Comment (JanniRoebbecke) [reply protzilla/data_preprocessing/filter_samples.py]: default values where removed [https://github.com/cschlaffner/PROTzilla2/pull/496/commits/c90e845d2d921d60472dc97ea459a0413058df88](url)

File: protzilla/data_preprocessing/filter_proteins.py [hunk 3]
```diff
@@ -4,14 +4,19 @@
...
     """
     This function filters proteins based on the amount of samples with nan values, if the percentage of nan values
     is below a threshold (percentage).
 
-    :param df: the intensity dataframe that should be filtered
+    :param protein_df: the protein dataframe that should be filtered
+    :param peptide_df: the peptide dataframe that should be filtered in accordance to the intensity dataframe (optional)
     :param percentage: ranging from 0 to 1. Defining the relative share of samples the proteins need to be present
-        in in order for the protein to be kept.
+        inorder for the protein to be kept.
```
[the remark this answers is not in the document]
Comment (JanniRoebbecke) [reply protzilla/data_preprocessing/filter_proteins.py]: [https://github.com/cschlaffner/PROTzilla2/pull/496/commits/c90e845d2d921d60472dc97ea459a0413058df88](url)

File: protzilla/data_preprocessing/filter_proteins.py [hunk 4]
```diff
@@ -25,9 +30,17 @@ def by_samples_missing(protein_df: pd.DataFrame, percentage: float) -> dict:
     filtered_proteins_list = (
         transformed_df.drop(remaining_proteins_list, axis=1).columns.unique().tolist()
     )
-    filtered_df = protein_df[(protein_df["Protein ID"].isin(remaining_proteins_list))]
+    filtered_df = protein_df[
+        (protein_df["Protein ID"].isin(remaining_proteins_list))
+    ]
+    filtered_peptide_df = None
+    if peptide_df is not None:
```
Comment (lucatreide) [inline protzilla/data_preprocessing/filter_proteins.py:37]: Kann man kürzer schreiben mit:
`filtered_peptide_df = peptide_df[...] if peptide_df else None`

File: protzilla/data_preprocessing/filter_samples.py [hunk 5]
```diff
@@ -27,18 +30,32 @@ def by_protein_intensity_sum(
             (median + deviation_threshold * sd),
         )
     ].index.tolist()
+
+    filtered_df = protein_df[~(protein_df["Sample"].isin(filtered_samples_list))]
+    filtered_peptide_df = None
```
Comment (lucatreide) [inline protzilla/data_preprocessing/filter_samples.py:35]: Gleiches wie bei `by_samples_missing`

File: protzilla/data_preprocessing/filter_samples.py [hunk 6]
```diff
@@ -59,18 +76,32 @@ def by_protein_count(protein_df: pd.DataFrame, deviation_threshold: float) -> di
...
```
Comment (lucatreide) [inline protzilla/data_preprocessing/filter_samples.py:81]: siehe `by_samples_missing`

File: protzilla/data_preprocessing/filter_samples.py [hunk 7]
```diff
@@ -87,8 +118,17 @@ def by_proteins_missing(protein_df: pd.DataFrame, percentage: float) -> dict:
     filtered_samples_list = sample_protein_count[
         ~sample_protein_count.ge(total_protein_count * percentage)
     ].index.tolist()
+
+    filtered_df = protein_df[~(protein_df["Sample"].isin(filtered_samples_list))]
+    filtered_peptide_df = None
```
Comment (lucatreide) [inline protzilla/data_preprocessing/filter_samples.py:123]: siehe `by_samples_missing`

File: protzilla/data_preprocessing/imputation.py [hunk 8]
```diff
@@ -57,6 +57,7 @@ def flag_invalid_values(df: pd.DataFrame, messages: list) -> dict:
 
 def by_knn(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:60]: Kann peptide_df None sein? Dann als Typen `pd.DataFrame | None `angeben

File: protzilla/data_preprocessing/imputation.py [hunk 9]
```diff
@@ -104,6 +105,7 @@ def by_knn(
 
 def by_simple_imputer(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:108]: siehe `by_knn`

File: protzilla/data_preprocessing/imputation.py [hunk 10]
```diff
@@ -144,6 +146,7 @@ def by_simple_imputer(
 
 def by_min_per_sample(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:149]: siehe `by_knn`

File: protzilla/data_preprocessing/imputation.py [hunk 11]
```diff
@@ -186,6 +189,7 @@ def by_min_per_sample(
 
 def by_min_per_protein(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:192]: siehe `by_knn`

File: protzilla/data_preprocessing/imputation.py [hunk 12]
```diff
@@ -229,6 +233,7 @@ def by_min_per_protein(
 
 def by_min_per_dataset(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:236]: siehe `by_knn`

File: protzilla/data_preprocessing/imputation.py [hunk 13]
```diff
@@ -259,6 +264,7 @@ def by_min_per_dataset(
 
 def by_normal_distribution_sampling(
     protein_df: pd.DataFrame,
+    peptide_df: pd.DataFrame,
```
Comment (lucatreide) [inline protzilla/data_preprocessing/imputation.py:267]: siehe `by_knn`

File: protzilla/data_preprocessing/imputation.py (hunk 8 above)
Comment (lucatreide) [reply protzilla/data_preprocessing/imputation.py:60]: Und warum wird das überhaupt übergeben, wenn wir das nicht benutzen?

File: protzilla/data_preprocessing/normalisation.py [hunk 14]
```diff
@@ -8,7 +8,7 @@
 from protzilla.utilities import default_intensity_column
 
 
-def by_z_score(protein_df: pd.DataFrame) -> dict:
+def by_z_score(protein_df: pd.DataFrame, peptide_df: pd.DataFrame) -> dict:
```
Comment (lucatreide) [inline protzilla/data_preprocessing/normalisation.py:11]: siehe `knn`

File: protzilla/data_preprocessing/outlier_detection.py [hunk 15]
```diff
@@ -59,9 +62,12 @@ def by_isolation_forest(
         ].index.tolist()
 
         protein_df = protein_df[~(protein_df["Sample"].isin(outlier_list))]
+        peptide_df = (None if peptide_df is None
```
Comment (lucatreide) [inline protzilla/data_preprocessing/outlier_detection.py:65]: Probably ist `peptide_df[~(peptide_df["Sample"].isin(outlier_list))] if peptide_df else None` lesbarer, da der "relevante" Part am Anfang steht.

File: protzilla/data_preprocessing/outlier_detection.py [hunk 16]
```diff
@@ -117,8 +125,12 @@ def by_local_outlier_factor(
         outlier_list = df_lof_data[df_lof_data["Outlier"]].index.tolist()
 
         protein_df = protein_df[~(protein_df["Sample"].isin(outlier_list))]
+        peptide_df = (None if peptide_df is None
```
Comment (lucatreide) [inline protzilla/data_preprocessing/outlier_detection.py:128]: siehe `by_isolation_forest`

File: protzilla/data_preprocessing/outlier_detection.py [hunk 17]
```diff
@@ -218,9 +232,12 @@ def by_pca(
             df_transformed_pca_data["Outlier"]
         ].index.tolist()
         protein_df = protein_df[~(protein_df["Sample"].isin(outlier_list))]
+        peptide_df = (None if peptide_df is None
```
Comment (lucatreide) [inline protzilla/data_preprocessing/outlier_detection.py:235]: siehe `by_isolation_forest`

File: ui/runs/forms/importing.py [hunk 18]
```diff
@@ -126,3 +126,34 @@ class PeptideImportForm(MethodForm):
...
+        choices=IntensityType, label="Intensity parameter"
+    )
+    map_to_uniprot = CustomBooleanField(
+        label="Map to Uniprot IDs using Biomart (online)", required=False
+    )
+
+    def fill_form(self, run: Run) -> None:
+        from protzilla.methods.importing import MaxQuantImport, MsFraggerImport, DiannImport
+
+        self.fields["intensity_name"].initial = (
```
Comment (lucatreide) [inline ui/runs/forms/importing.py:155]: Do the initials make sense? Is it possible to change if you want to `map_to_uniprot` between protein and peptide import?

**Review by lucatreide — APPROVED** [review]
Lgtm, small things

**Pull request merged**